Fix and modularize skill get expected damage

LordMidas opened the pull request:
This is a draft PR to get your first impressions and feedback/ideas on how to approach this. 

This PR intends to completely rewrite the `getExpectedDamage` logic to utilize modularized DRY functions and also to make the logic be more accurate. The vanilla implementation has several issues including but not limited to:
- Using wrong properties (`buildPropertiesForDefense` instead of `buildPropertiesForBeingHit`).
- Not accounting for many vanilla damage related fields e.g. `skill.m.DamageMinimum` and `properties.DamageRegularReduction` and others.

Related discussion on Discord: https://discord.com/channels/965324395851694140/1000342646184747008/1367125435070287995

LordMidas committed:
feat: add function to init HitInfo

LordMidas committed:
feat: initialize properties in getDamage functions in skill if null

And make it an optional parameter that defaults to null.

Suor commented on mod_modular_vanilla/hooks/skills/skill.nut:
```diff
@@ -6,11 +6,85 @@
+		}
+
+		local headshotChance = p.getHitchance(::Const.BodyPart.Head)
+		if (headshotChance != 0 && !_target.getCurrentProperties().IsImmuneToCriticals && !_target.getCurrentProperties().IsImmuneToHeadshots)
```
So if target is immune to head shots then we do not calc damage towards head at all? Looks wrong. These checks should probably be moved elsewhere.

LordMidas replied:
Fixed. I moved the check up and set headshot chance to 0 if target is immune, so we calculate based on body shot rules.

Suor reviewed (changes requested):
Also, 2H axes split man hitting both head and body not considered here. Maybe not here but should go somewhere.

LordMidas committed:
fix: set bodyshot chance to 100 if target is immune to headshots

LordMidas committed:
refactor: set DirectDamage to 0 and add comment about calculation

LordMidas committed:
docs: add comments

Suor commented on mod_modular_vanilla/hooks/skills/skill.nut:
```diff
@@ -6,11 +6,85 @@
+		local p = this.getContainer().buildPropertiesForUse(this, _target);
+
+		// Set the damage in the properties to the average damage so that our MV_getDamageXYZ functions always roll the average damage
+		local damageRegularAvg = ::Math.floor((p.DamageRegularMin + p.DamageRegularMax) * 0.5);
+		p.DamageRegularMin = damageRegularAvg;
+		p.DamageRegularMax = damageRegularAvg;
```
I mean it will make wrong estimations, you should not use average in the presence of reductions. I guess vanilla does it wrong too.

Example: if target has big reductions then a skill with bigger damage spread is preferred, this is lost here.

LordMidas replied:
I see. Perhaps for "most AI" calculations probably using average is ok in favor of efficiency. But if we want to do it more intelligent like you suggest then what would be the approach?
- First roll min damage and calculate HP damage received
- Then roll max damage and calculate HP damage received
- Then calculate average of the above two?
That will increase the operations by 2x. I already increased the operations compared to vanilla by 2x, so if we do this more intelligent version now it will be 4x more operations than vanilla.

Suor replied:
Probably leave this for later

Suor commented on mod_modular_vanilla/hooks/skills/skill.nut:
```diff
@@ -6,11 +6,107 @@
+		local headshotChance = p.getHitchance(::Const.BodyPart.Head);
+		if (_target.getCurrentProperties().IsImmuneToCriticals || _target.getCurrentProperties().IsImmuneToHeadshots)
+		{
+			headshotChance = 0;
```
I don't think this is how it works. Even if there is no crit on head I still might want to hit there because of low armor

LordMidas replied:
I see, yes that's true. Unfortunately, vanilla did some really dirty work here. They check for these properties inside `actor.onDamageReceived` and manually set the `hitInfo.BodyDamageMult = 1.0` there. So, when "predicting" damage in this `getExpectedDamage` function we would also have to manually do that. I don't like this because if someone changes that part then he also has to remember to update this function to change that. I hate this part of vanilla where things are coded in a non-modular and out of place way.

So what do you propose as the best solution here? Probably something like:

```squirrel
// This property in vanilla is for entities who are not meant to have any head, so they don't receive any headshots
if (_target.getCurrentProperties().IsImmuneToHeadshots)
{
   headshotChance = 0;
}
```

And then later where we build hitInfo for headshot part, we hardcode it as vanilla does in `actor.onDamageReceived`?
```squirrel
if (_target.getCurrentProperties().IsImmuneToCriticals)
{
   hitInfo.BodyDamageMult = 1.0;
}
```

LordMidas replied:
Oh my God, I just looked inside vanilla code and have seen that vanilla doesn't even redirect all hits to body even if target is `ImmuneToHeadshots`..... this is very disappointing. So I suppose we have to do the following inside our headshot part just like vanilla does it in `actor.onDamageReceived` even though I hate it:
```squirrel
if (this.m.CurrentProperties.IsImmuneToCriticals || this.m.CurrentProperties.IsImmuneToHeadshots)
{
	hitInfo.BodyDamageMult = 1.0;
}
```

LordMidas committed:
fix: manually set BodyDamageMult to 1.0 just like actor.onDamageReceived

LordMidas committed:
fix: arg order in call to MV_initHitInfo

LordMidas committed:
fix: add properties for defense and being hit to hitInfo MV fields

LordMidas committed:
perf: calculate ratio mult once

LordMidas committed:
refactor: move initHitInfo function to Const.Tactical

LordMidas committed:
refactor: remove redundant armor calculation

Vanilla needed it because they calculate use it to calculate the HP damage mitigation from armor but we don't need it because we use our MV functions to calculate damage.

LordMidas committed:
feat: fix and modularize vanilla getExpectedDamage

LordMidas committed:
docs: add comment about split man not being accounted

LordMidas committed:
fix: body damage mult not being properly forced to head or body

LordMidas committed:
refactor: improve comment about onBeforeTargetHit

LordMidas committed:
fix: bodyPart var missing in onScheduledTargetHit after modularization

LordMidas committed:
docs: improve comment about onBeforeTargetHit

LordMidas committed:
fix: adjust the HitChanceMult accordingly to force head or body hit

LordMidas committed:
feat: account for three headed flail attacks higher max damage potential

Darxo commented on mod_modular_vanilla/hooks/skills/skill.nut:
```diff
@@ -1,9 +1,119 @@
+			local hitInfo = ::Const.Tactical.MV_initHitInfo(this, _target, p, d);
+
+			// This will now use the outgoing hitInfo to prepare the correct properties for receiving damage
+			hitInfo.MV_PropertiesForBeingHit = _target.getSkills().buildPropertiesForBeingHit(actor, this, hitInfo);
```
Since `buildPropertiesForBeingHit` triggers `onBeforeDamageReceived` for all skills, this might have permanent sideeffects.

For example `bone_platings_upgrade` is depleting its charge during `onBeforeDamageReceived`. Though looking at the Vanilla code, the `onBeforeDamageReceived` of items is not called during `buildPropertiesForBeingHit`; that event is only triggered for skills there

LordMidas replied:
It is indeed possible that `onBeforeDamageReceived` triggers some permanent side-effects e.g. increasing/decreasing stacks of some perks/skills. Ideally, we shouldn't have to do this if vanilla implemented proper separation between functions that modify properties vs functions that modify `hitInfo` (i.e. an _actual_ hit). 

As it stands, vanilla has implemented damage reduction for racial skills (and nimble / battle forged) in `onBeforeDamageReceived`. Ideally it should have been on `onBeingAttacked` which is a non-destructive `properties`-generating function.

But in order to cover the vanilla racial skills (and any modded skills that use that function in vanilla style for damage reduction) we have to call this here. I don't see another way.

Generally speaking, best practice would be to consider `buildPropertiesForXYZ` functions to be "non-state-changing" and never use them to trigger permanent things.

Darxo replied:
Perhaps we implement new `preview`-versions of those skill events/getter.
And when we call those, they flip a `IsPreviewing`/`IsSimulating` global variable, which our perks can then check for

LordMidas replied:
But that way we will not cover mods that don't use our preview type functions. So, effects of those mods will not be accounted for.

But yes, calling this function can indeed cause some mods to trigger effects because they don't expect this function to be called except during `actor.onDamageReceived`. I just noticed that in Reforged we have many skills/effects that trigger things during `onBeforeDamageReceived`.

Btw, I also realized that there is already a way for us to check whether this is an actual attack or a preview check. E.g. look in `rf_whimpering_veil_effect` in Reforged. I use a check for `Tactical.State.MV_getCurrentAttackInfo() == null` as a guard to ensure that we are inside an actual attack. But the issue is that other mods cannot be expected to use this as most mods are not built with Modular Vanilla in mind.

LordMidas replied:
An alternative route is that we fix vanilla skills to apply their damage reduction during `onBeingAttacked` i.e. `buildPropertiesForDefense` instead. Then we don't call `buildPropertiesForBeingHit` here at all.

But that will also suffer from the problem of not incorporating damage reduction from mods which may be using the vanilla style of applying them in `onBeforeDamageReceived`.

LordMidas commented on mod_modular_vanilla/hooks/skills/skill.nut:
```diff
@@ -1,9 +1,119 @@
+			// This will now use the outgoing hitInfo to prepare the correct properties for receiving damage
+			hitInfo.MV_PropertiesForBeingHit = _target.getSkills().buildPropertiesForBeingHit(actor, this, hitInfo);
+
+			// Vanilla changes the HitInfo in certain skills in onBeforeTargetHit e.g. `pound` and `gash_skill`.
+			// Vanilla also spawns icon for `perk_coup_de_grace` in `onBeforeTargetHit`.
+			// Some mods may do some state changes in `onBeforeTargetHit`.
+			// TODO: Therefore I am not sure how we can account for these. Calling this function doesn't seem safe.
+			// local wasUpdating = this.getContainer().m.IsUpdating;
+			// this.getContainer().m.IsUpdating = true;
+			// this.getContainer().onBeforeTargetHit(this, _target, hitInfo);
+			// this.getContainer().m.IsUpdating = wasUpdating;
```
They are only relevant to the commented out code.
